### PR TITLE
Detailed Balance loss

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**
This PR implements the detailed balance objective, building upon some of the code developed for the forward-looking loss.

Experiments with the Grid and the Tetris on wandb: https://wandb.ai/alexhg/detailedbalance

### CODE REVIEW

File: gflownet/gflownet.py [hunk 1]
```diff
@@ -679,6 +682,65 @@ def trajectorybalance_loss(self, it, batch):
...
+        logflow_states = self.state_flow(states_policy).squeeze()
+        logflow_states[done.eq(1)] = torch.log(rewards)
+        # TODO: Optimise by reusing logflow_states and batch.get_parent_indices
+        logflow_parents = self.state_flow(parents_policy).squeeze()
+
+        # Detailed balance loss
+        loss = (
+            (logflow_parents + logprobs_f - logflow_states - logprobs_b).pow(2).mean()
+        )
+        return loss, loss, loss
```
Comment (AlexandraVolokhova) [inline gflownet/gflownet.py:742]: I think here it is possible to distinguish between terminal and non-terminal losses, so I'd do that instead of returning the full loss tree times

**Review by AlexandraVolokhova — APPROVED** [review]

File: gflownet/gflownet.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply gflownet/gflownet.py:742]: ooooook I'll do it

Comment (alexhernandezgarcia) [reply gflownet/gflownet.py]: https://github.com/alexhernandezgarcia/gflownet/pull/256/commits/293010f29fb754404587f0d8f60091bd2d0c43c0
